### PR TITLE
Victor/core/header with empty value

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-c/issues/621

Required to support adding a header with no value

### CODE REVIEW

File: sdk/core/core/src/az_http_request.c [hunk 1]
```diff
@@ -148,7 +148,7 @@ az_http_request_append_header(_az_http_request* p_request, az_span key, az_span
 {
   AZ_PRECONDITION_NOT_NULL(p_request);
   AZ_PRECONDITION_VALID_SPAN(key, 1, false);
-  AZ_PRECONDITION_VALID_SPAN(value, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(value, 0, true);
 
   AZ_PRECONDITION(az_span_size(key) > 0 || az_span_size(value) > 0);
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_request.c:153]: I think we don't need this precondition anymore.

File: sdk/core/core/src/az_span.c [hunk 2]
```diff
@@ -234,7 +234,7 @@ az_span az_span_copy(az_span destination, az_span source)
   // Implementations of memmove generally do the right thing when number of bytes to move is 0, even
   // if the ptr is null, but given the behavior is documented to be undefined, we disallow it as a
   // precondition.
-  AZ_PRECONDITION_VALID_SPAN(source, 0, false);
+  AZ_PRECONDITION_VALID_SPAN(source, 0, true);
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_span.c:237]: Please revert this change. This check should be made by the caller and they should not pass in `NULL` ptr to `az_span_copy`. See the comment right above this line.

The fix doesn't belong here but rather in `az_http_request_append_header`.

File: sdk/core/core/src/az_http_request.c [hunk 3]
```diff
@@ -188,8 +188,17 @@ AZ_NODISCARD az_result az_http_request_get_parts(
     az_span* out_url,
     az_span* out_body)
 {
-  *out_method = request->_internal.method;
-  *out_url = az_span_slice(request->_internal.url, 0, request->_internal.url_length);
-  *out_body = request->_internal.body;
+  if (out_method != NULL)
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_request.c:191]: Why are we allowing null ptrs here rather than these being preconditions? At the very least this seems unrelated to the fix itself.

File: sdk/core/core/src/az_http_request.c [hunk 4]
```diff
@@ -148,7 +148,7 @@ az_http_request_append_header(_az_http_request* p_request, az_span key, az_span
 {
   AZ_PRECONDITION_NOT_NULL(p_request);
   AZ_PRECONDITION_VALID_SPAN(key, 1, false);
-  AZ_PRECONDITION_VALID_SPAN(value, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(value, 0, true);
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_request.c:151]: For my edification, can you please provide some context on why this should be allowed? What is motivating this change?

File: sdk/core/core/src/az_span.c (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_span.c:237]: well, this was @JeffreyRichter request actually
Since we should let customers call az_span_copy with NULL as sorce and that would just return dst without doing anything else.

File: sdk/core/core/src/az_http_request.c (hunk 3 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_request.c:191]: not for the issue, right
This is part of a general problem where transport adapter is accesing internal fields.
I can remove this from this PR and add it later.
This is to allow calling get parts just to get method with out worrying about headers and body from a request

File: sdk/core/core/src/az_http_request.c (hunk 4 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_request.c:151]: headers with no value are valid on http requests.
See the issue description. Current precondition block us from adding  a header with no value.
See the unit tests I added as well

File: sdk/core/core/src/az_http_request.c (hunk 3 above)
Comment (ahsonkhan) [reply sdk/core/core/src/az_http_request.c:191]: Oh I see. So is this so that the caller can pick which part they are interested in and pass in `NULL` for the parts they don't need?

Comment (ahsonkhan) [reply sdk/core/core/src/az_http_request.c:191]: Please update the docs for `az_http_request_get_parts` to mention this.

File: sdk/core/core/src/az_http_request.c (hunk 4 above)
Comment (ahsonkhan) [reply sdk/core/core/src/az_http_request.c:151]: > headers with no value are valid on http requests.

Is there a spec or doc that I can read about that?

I found this but it doesn't mention the value can be optional.
https://developer.mozilla.org/en-US/docs/Glossary/Request_header

File: sdk/core/core/src/az_span.c (hunk 2 above)
Comment (ahsonkhan) [reply sdk/core/core/src/az_span.c:237]: I agree that copying an **empty** span as the source into a destination should just return dst without doing anything else (for example a slice of a valid buffer that happens to have size 0). I am not sure that **a span with a null ptr** should be allowed to be passed in to `az_span_copy` though (such as `AZ_SPAN_NULL`).

And if we really want to support that here, in az_span_copy, we need to also add a check in `az_span_copy` to handle null ptr OR update the comment that warns against passing in null to `memmove`:

```C
uint8_t* ptr = az_span_ptr(source);
if (ptr == NULL)
{
   return destination;
}
```

File: sdk/core/core/test/cmocka/test_az_http.c [hunk 5]
```diff
@@ -145,6 +147,31 @@ static void test_http_request(void** state)
...
+        &az_context_app,
+        az_http_method_get(),
+        url_span,
+        az_span_size(hrb_url),
+        header_span,
+        AZ_SPAN_FROM_STR("body")));
+
+    // Empty header
+    assert_return_code(
+        az_http_request_append_header(&hrb, AZ_SPAN_FROM_STR("header"), AZ_SPAN_NULL), AZ_OK);
```
Comment (ahsonkhan) [inline sdk/core/core/test/cmocka/test_az_http.c:173]: Since we are now allowing `AZ_SPAN_NULL` as the value in the key:value `az_pair`, do we need more changes in `az_http_request_append_header` on the internal representation using `az_pair` and how we increase the offset? Should we not use `az_pair` anymore for this if one of the spans are optional? I don't know if increasing by `sizeof(az_pair)` is correct if we are only adding a single header span, with no value. Can you make sure the offsets are done correctly?

File: sdk/core/core/src/az_http_request.c (hunk 4 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_request.c:151]: this is the RFC spec: https://tools.ietf.org/html/rfc7230#section-3.2

File: sdk/core/core/src/az_span.c (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_span.c:237]: I will add that if condition.
We need to support it because this can happen on runtime (where preconditions have might been compiled away)
Imagine this
```c
az_span some_span = {0};
// Here customer call some function from their side to read/get some az_span from I/O
some_span = customer_funct(io);  // Even if this is consider to never fail, we can't be sure `some_span` will be updated to point somewhere..

// then customer try to copy 
az_span_copy(other_buffer, some_span);  // we can't tell customer to check that `some_span` is not null, because it can be runtime-calculated
```
 
```

File: sdk/core/core/test/cmocka/test_az_http.c (hunk 5 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_http.c:173]: yeah, we still need az_pair.  It is the same size to have  one `az_span` or `AZ_SPAN_NULL`.  And the trouble of managing not adding the value when it is null is high!  compared to what we would win.
For instance, for looping headers, we would need a separator to know where each header end (instead of fixed size of az_pair)

File: sdk/core/core/inc/az_http_transport.h [hunk 6]
```diff
@@ -85,12 +85,16 @@ AZ_NODISCARD az_result
 az_http_request_get_header(_az_http_request const* request, int32_t index, az_pair* out_header);
 
 /**
- * @brief Get parts of an HTTP request.
+ * @brief Get parts of an HTTP request. NULL in accepted to ignore getting any parts, for example,
+ * call this function like below to get only the http method and ignore getting url and body
+ *   az_http_request_get_parts(request, &method, NULL, NULL)`
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_http_transport.h:90]: You are missing a back tick in the beginning of this line of code. Can you check how this renders in doxygen?

Also, can you surround NULL in all the places within this APIs docs with back ticks, i.e. `NULL`?

File: sdk/core/core/inc/az_http_transport.h [hunk 7]
```diff
@@ -85,12 +85,16 @@ AZ_NODISCARD az_result
 az_http_request_get_header(_az_http_request const* request, int32_t index, az_pair* out_header);
 
 /**
- * @brief Get parts of an HTTP request.
+ * @brief Get parts of an HTTP request. NULL in accepted to ignore getting any parts, for example,
+ * call this function like below to get only the http method and ignore getting url and body
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_http_transport.h:89]:
```suggestion
 * call this function like below to get only the http method and ignore getting url and body.
```

Also consider moving the 2nd sentence with example to a `@note` or `@remarks` instead of making it part of the `@brief`.

File: sdk/core/core/inc/az_http_transport.h [hunk 8]
```diff
@@ -85,12 +85,16 @@ AZ_NODISCARD az_result
...
+ *   az_http_request_get_parts(request, &method, NULL, NULL)`
+ *
+ * This function is expected to be used by transport layer only.
  *
  * @param request HTTP request to get parts from.
- * @param out_method Pointer to write HTTP method to.
- * @param out_url Pointer to write URL to.
- * @param out_body Pointer to write HTTP request body to.
+ * @param out_method Pointer to write HTTP method to. Use NULL to ignore getting this value.
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_http_transport.h:95]: Annotate the parameter as nullable.
See:
https://github.com/Azure/azure-sdk-for-c/blob/da98d45fb4a1d44b940cec3e69a9358d566879c7/sdk/iot/provisioning/inc/az_iot_provisioning_client.h#L68-L70
https://azure.github.io/azure-sdk/clang_documentation.html#docstrings

File: sdk/core/core/inc/az_span.h [hunk 9]
```diff
@@ -252,6 +252,8 @@ AZ_NODISCARD int32_t az_span_find(az_span source, az_span target);
  * @remarks The method assumes that the \p destination has a large enough size to hold the \p
  * source.
  * @remarks This method copies all of \p source into the \p destination even if they overlap.
+ * @remarks If \p source is an empty #az_span or #AZ_SPAN_NULL, this function will just return
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_span.h:255]: Nice!

File: sdk/core/core/src/az_http_request.c [hunk 10]
```diff
@@ -182,14 +179,37 @@ az_http_request_get_header(_az_http_request const* request, int32_t index, az_pa
...
+ *
+ * This function is expected to be used by transport layer only.
+ *
+ * @param request HTTP request to get parts from.
+ * @param out_method Pointer to write HTTP method to. Use NULL to ignore getting this value.
+ * @param out_url Pointer to write URL to. Use NULL to ignore getting this value.
+ * @param out_body Pointer to write HTTP request body to. Use NULL to ignore getting this value.
+ *
+ * @retval AZ_OK Success.
+ */
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_request.c:195]: We don't need to doc the *.c file if the API is already exposed from the *.h file and documented there.

File: sdk/core/core/inc/az_http_transport.h [hunk 11]
```diff
@@ -85,12 +85,16 @@ AZ_NODISCARD az_result
...
+ * @param out_url Pointer to write URL to. Use NULL to ignore getting this value.
+ * @param out_body Pointer to write HTTP request body to. Use NULL to ignore getting this value.
  *
  * @retval AZ_OK Success.
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_http_transport.h:99]: We should try to be consistent with our docs, especially those returning az_result.
1) Consider using `@return`
2) Consider formatting it as a list, like this:
https://github.com/Azure/azure-sdk-for-c/blob/da98d45fb4a1d44b940cec3e69a9358d566879c7/sdk/core/core/inc/az_span.h#L301-L304

File: sdk/core/core/src/az_span.c [hunk 12]
```diff
@@ -231,15 +231,15 @@ AZ_NODISCARD int32_t az_span_find(az_span source, az_span target)
...
-  // Implementations of memmove generally do the right thing when number of bytes to move is 0, even
-  // if the ptr is null, but given the behavior is documented to be undefined, we disallow it as a
-  // precondition.
-  AZ_PRECONDITION_VALID_SPAN(source, 0, false);
-
   int32_t src_size = az_span_size(source);
 
   AZ_PRECONDITION_VALID_SPAN(destination, src_size, false);
 
+  if (src_size == 0 || az_span_ptr(source) == NULL)
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_span.c:238]: I don't think we need the `az_span_ptr(source) == NULL` check.

`if (src_size == 0)` should be enough to capture both conditions.

Also, please keep/modify the comment and move it right above this line, since it captures why we are even doing the check and not just letting it go through to memmove.

File: sdk/core/core/test/cmocka/test_az_span.c [hunk 13]
```diff
@@ -751,6 +751,14 @@ static void az_span_u32toa_overflow_fails(void** state)
   assert_true(az_span_u32toa(buffer, v, &out_span) == AZ_ERROR_INSUFFICIENT_SPAN_SIZE);
 }
 
+static void az_span_copy_empty(void** state)
```
Comment (ahsonkhan) [inline sdk/core/core/test/cmocka/test_az_span.c:754]: While you are adding tests here, can you please add cases where both source and destination are empty as well.

File: sdk/core/core/src/az_span.c (hunk 2 above)
Comment (ahsonkhan) [reply sdk/core/core/src/az_span.c]: Couldn't the same be said about the destination? Why is destination not allowed to be `AZ_SPAN_NULL`?

File: sdk/core/core/inc/az_http_transport.h (hunk 11 above)
Comment (antkmsft) [reply sdk/core/core/inc/az_http_transport.h:99]: BTW, I think we should use `@retval` everywhere we use `@return` as in example above.

Comment (ahsonkhan) [reply sdk/core/core/inc/az_http_transport.h:99]: I am curious to know, why you prefer one over the other? Does it render differently?

Either way, let's try to be consistent.

File: sdk/core/core/src/az_span.c (hunk 12 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_span.c:238]: fine, I can remove it.
But this goes against what you said before about the more we validate weird cases, the best shape we are XD

File: sdk/core/core/test/cmocka/test_az_span.c (hunk 13 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_span.c:754]: not relevant to this PR. 
That should be part of a different issue that would cover lack of coverage for az_span_copy, of something like that.

here we only care about managing empty source without considering destination

File: sdk/core/core/inc/az_http_transport.h (hunk 11 above)
Comment (antkmsft) [reply sdk/core/core/inc/az_http_transport.h:102]: `@retval`'s purpose is exactly to specify each concrete return value: i.e. for `az_result convert_roman_numeral_to_decimal(az_span roman, az_span* out_decimal)`, it can be "`AZ_OK` success, `AZ_ERROR_ARG` input is malformed, `AZ_ERROR_INSUFFICIENT_SPAN_CAPACITY` not enough size to fit the result".

`@return`'s goal is do describe the return value in general, i.e.  for `int strlen(char const* s)` it can be `string length, -1 if s is NULL`.

For the cases when we describe each return value, we happen to format `return` ourselves to describe the return the values, and we may end up matching `@retval`'s output pretty close, but this is what `@retval` is for - it will use Doxygen's method to format them. But anyways, it is readable, but I think ideally it is better to use the specific tool for the job.

File: sdk/core/core/src/az_span.c (hunk 12 above)
Comment (antkmsft) [reply sdk/core/core/src/az_span.c]: And what about negative source length? :) Is it a  `precondition(source, 0, true)` then, or the `if` should be `size <= 0` instead of just `== 0`

File: sdk/core/core/inc/az_http_transport.h [hunk 14]
```diff
@@ -85,16 +85,19 @@ AZ_NODISCARD az_result
 az_http_request_get_header(_az_http_request const* request, int32_t index, az_pair* out_header);
 
 /**
- * @brief Get parts of an HTTP request. NULL in accepted to ignore getting any parts, for example,
- * call this function like below to get only the http method and ignore getting url and body
- *   az_http_request_get_parts(request, &method, NULL, NULL)`
+ * @brief Get parts of an HTTP request. `NULL` in accepted to ignore getting any parts, for example,
+ * call this function like below to get only the http method and ignore getting url and body.
+ *   `az_http_request_get_parts(request, &method, `NULL`, `NULL`)`
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_http_transport.h:90]: Since this is a line of code, you don't need back ticks around the `NULL` parameters. Also, I don't know how embedded back ticks renders.
I would just keep this:
`az_http_request_get_parts(request, &method, NULL, NULL)`

File: sdk/core/core/inc/az_http_transport.h [hunk 15]
```diff
@@ -85,16 +85,19 @@ AZ_NODISCARD az_result
...
  *
  * This function is expected to be used by transport layer only.
  *
  * @param request HTTP request to get parts from.
- * @param out_method Pointer to write HTTP method to. Use NULL to ignore getting this value.
- * @param out_url Pointer to write URL to. Use NULL to ignore getting this value.
- * @param out_body Pointer to write HTTP request body to. Use NULL to ignore getting this value.
+ * @param out_method[out] __[nullable]__ Pointer to write HTTP method to. Use `NULL` to ignore
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_http_transport.h:95]: The direction [in]/[out] goes right after `@param`.

It should follow this format (from the design guidelines):
`@param[<direction>] <param_name> __<additional_annotations>__ description`

File: sdk/core/core/inc/az_http_transport.h (hunk 11 above)
Comment (ahsonkhan) [reply sdk/core/core/inc/az_http_transport.h:102]: Interesting. I will take a closer look. Thanks @antkmsft

File: sdk/core/core/inc/az_http_transport.h (hunk 15 above)
Comment (vhvb1989) [reply sdk/core/core/inc/az_http_transport.h:95]: please feel free to make any change on documentation directly into my branches XD lol.

File: sdk/core/core/src/az_span.c (hunk 12 above)
Comment (ahsonkhan) [reply sdk/core/core/src/az_span.c]: > But this goes against what you said before about the more we validate weird cases, the best shape we are XD

That was about precondition checks (where we are less cost sensitive and being extra cautious to catch dev bugs doesn't hurt), compared to checks in release code which has preconditions turned off, where perf/code size is at a premium. Minimizing and optimizing validation checks especially when trying to make trade offs, makes sense to me.

> And what about negative source length? :) Is it a  `precondition(source, 0, true)` then, or the `if` should be `size <= 0` instead of just `== 0`

I think both of those are good ideas :) (where I prefer the precondition and then `== 0` size check).

File: sdk/core/core/inc/az_http_transport.h [hunk 16]
```diff
@@ -85,12 +85,19 @@ AZ_NODISCARD az_result
 az_http_request_get_header(_az_http_request const* request, int32_t index, az_pair* out_header);
 
 /**
- * @brief Get parts of an HTTP request.
+ * @brief Get parts of an HTTP request. `NULL` in accepted to ignore getting any parts, for example,
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_http_transport.h:88]: > `NULL` in accepted to ignore getting any parts,

Should the `in` be `is`?

```suggestion
 * @brief Get parts of an HTTP request. `NULL` is accepted to ignore getting any parts, for example,
```

File: sdk/core/core/src/az_span.c [hunk 17]
```diff
@@ -231,15 +231,15 @@ AZ_NODISCARD int32_t az_span_find(az_span source, az_span target)
 
 az_span az_span_copy(az_span destination, az_span source)
 {
-  // Implementations of memmove generally do the right thing when number of bytes to move is 0, even
-  // if the ptr is null, but given the behavior is documented to be undefined, we disallow it as a
-  // precondition.
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_span.c:236]: It's unfortunate that we got rid of this comment :(

File: sdk/core/core/inc/az_http_transport.h (hunk 11 above)
Comment (ahsonkhan) [reply sdk/core/core/inc/az_http_transport.h:102]: Regarding `@retval` vs `@return`, retval does look pretty nice:
image
```C
/**
 * @brief Appends the beginning of a JSON object (i.e. `{`).
 *
 * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
 * the start of object to.
 *
 * @return An #az_result value indicating the result of the operation:
 *         - #AZ_OK if object start was appended successfully
 *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
 *         - #AZ_ERROR_JSON_NESTING_OVERFLOW if the depth of the JSON exceeds the maximum allowed
 *           depth of 64
 */
AZ_NODISCARD az_result az_json_builder_append_begin_object(az_json_builder* json_builder);

/**
 * @brief Appends the beginning of a JSON array (i.e. `[`).
 *
 * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
 * the start of array to.
 *
 * @retval #AZ_OK if array start was appended successfully
 * @retval #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
 * @retval #AZ_ERROR_JSON_NESTING_OVERFLOW if the depth of the JSON exceeds the maximum allowed
 * depth of 64
 */
AZ_NODISCARD az_result az_json_builder_append_begin_array(az_json_builder* json_builder);
```

Maybe we combine return and retval, and have both?
image
